### PR TITLE
Compatibility with Spring Boot 3

### PR DESCRIPTION
**Pull request by anthonydahanne**
Original work from @abelsromero at https://github.com/spring-cloud/spring-cloud-bindings/pull/89

Reworked to comply with [Option 2 suggested](https://github.com/spring-cloud/spring-cloud-bindings/pull/89#issuecomment-1370306843) here .

### CODE REVIEW

File: pom.xml
Comment (abelsromero) [inline pom.xml:172]: Maybe a left over from local testing?

**Comment by abelsromero** [comment]
Seems good to me 😄 My only comment is about a weird `<skip>true</skip>`

File: pom.xml
Comment (anthonydahanne) [reply pom.xml]: good catch!
yes, I disabled GPG signing when building locally
fixed!

**Comment by anthonydahanne** [comment]
Hello @Kehrlann !
> That's a good first step.
Well, the second actually, after @abelsromero !

> BindingsPropertiesProcessor#process && Semver updates

`BindingsPropertiesProcessor` seemed like public interface and I was breaking it introducing a new argument; hence the bump to 2.0.0 - I still think there could be downstream consumers adding their own Processors and they would be broken after this interface change

While we could, like you did in your patch, not expose the spring version in the `process` signature, it will make unit testing slightly less ideal: in your tests you used static mocking; something I wish we could avoid using simple and testable interfaces.

> Spring boot version

OK for Spring boot 2.5 (your patch used 2.5 even you mentioned 2.4 in your comment - that's fine I guess since 2.4 support ends in 10 days) - for some reason I thought 2.3 was supported until the end of the year.


> adding duplicate spring boot 3 classes for each processor

Well, I personally prefer the basic `if` statements; easier to read, less code to duplicate (Redis is an example where no common properties are shared; the mast majority of processors share properties between 2 and 3)

We could discuss this point together, but for now I can't see the value of updating the `spring.factories` with 20 duplicate classes

> integration tests

Yes, that's a very good point. 

So I agree that we'd need to have spring boot 2.x and 3.x projects being built with this library, and then we'd run them with a few bindings on the file system, and then we'd ask the actuator about properties and assert the bindings are properly picked up

I think the best tool for the job would be a 2 steps in the CI, 1 named "Spring Boot 2 project Integration Test" and the other named "Spring Boot 3 project Integration Test"

Each of those steps would be a some basic bash script such as (pseudo code):

```
cd src/test/resources/spring-boot-2
export SERVICE_BINDING_ROOT=$PWD/bindings && export JAVA_TOOL_OPTIONS="-Dorg.springframework.cloud.bindings.boot.enable=true"
./mvnw spring-boot:run &
assert $(curl localhost:8080/actuator/configprops | jq -r .spring.redis.client-name) == my-name || exit 1
```
Which I believe would replicate what a user would do - and even document how to use the library (took me some time to find out about the JAVA_TOOL_OPTIONS ...)

Now, I wish the CI would be in GH action for everyone to see and for at least getting PR feedback!

I actually volunteer to port the CI to GH actions, I quickly mentioned it to @pivotal-david-osullivan , I don't know if everybody would agree on this?

Would we address CI and integration tests before resuming work on this PR ?

**Comment by anthonydahanne** [comment]
we had a chat with @Kehrlann and we agreed on the following:

* we don't change the `process` signature, this library can be injected without user knowledge, we don't want to break users - we'll only bump the minor version
* instead we inject a `SpringBootVersionResolver` class responsible to find out the spring boot version at hand
* (7) additional processors for Spring Boot 3 will be added, preferably in the same class file
* integration tests will be added, preferably runnable by default via maven

**Comment by anthonydahanne** [comment]
hello @Kehrlann 
Here's the updated version, you can have a look already; integration tests will be sent in another commit
Thanks!

File: spring-cloud-bindings/src/main/java/org/springframework/cloud/bindings/boot/SpringBootVersionResolver.java
Comment (Kehrlann) [inline spring-cloud-bindings/src/main/java/org/springframework/cloud/bindings/boot/SpringBootVersionResolver.java:23]: **[issue]** Make this package-protected ; you don't want this class to be a part of your public API (same as `MapMapper`).
**[suggestion]** set the default value for `forcedVersion` to `-1`, it's a usual convention for "this is invalid".
**[suggestion]** Maybe add a comment, "this class is used in unit tests, to fake the resolution of which major version of spring boot is used" or something similar.

```suggestion
class SpringBootVersionResolver {

    private int forcedVersion = -1;

    SpringBootVersionResolver() {
    }

    protected SpringBootVersionResolver(int forcedVersion) {
        this.forcedVersion = forcedVersion;
    }

   protected boolean isBootMajorVersionEnabled(int bootVersion) {
        if (forcedVersion != -1) {
            return forcedVersion == bootVersion;
        }
        int major = Integer.parseInt(SpringBootVersion.getVersion().split("\\.")[0]);
        return major == bootVersion;
    }
}
```

**Review by Kehrlann — COMMENTED** [review]
Great work, nice testing projects! 👏 

Comments: One issue re: public api, and a few suggestions. Two important points below:

### Testing suggestions

In the test projects, consider moving the `bindings` directories in the `src/test/resources` folder and reading them to select the Redis port. See [this patch](https://github.com/spring-cloud/spring-cloud-bindings/files/10756281/patch.txt)

### Major version upgrade?

We are introducing a breaking change here, e.g. `RedisBindingsPropertiesProcessor` is no longer a `BindingsPropertiesProcessor`.

Two choices: bump the major version ; or make `RedisBindingsPropertiesProcessor` behave like `RedisBindingsPropertiesProcessor.Boot2`. I think in that case the API doesn't change, only final classes, so in this case it's not the most disruptive change. I'll let you decide.

File: spring-cloud-bindings-tests-boot-2/src/test/java/org/springframework/cloud/SpringBoot2Application.java
Comment (anthonydahanne) [reply spring-cloud-bindings-tests-boot-2/src/test/java/org/springframework/cloud/SpringBoot2Application.java:9]: ok

File: spring-cloud-bindings-tests-boot-2/src/test/java/org/springframework/cloud/SpringBoot2Application.java
Comment (anthonydahanne) [reply spring-cloud-bindings-tests-boot-2/src/test/java/org/springframework/cloud/SpringBoot2Application.java:8]: ok

File: spring-cloud-bindings-tests-boot-2/pom.xml
Comment (anthonydahanne) [reply spring-cloud-bindings-tests-boot-2/pom.xml:24]: ok

**Comment by anthonydahanne** [comment]
All set! thanks for your reviews @abelsromero and @Kehrlann !

**Review by abelsromero — APPROVED** [review]
I can approve if necessary, but (disclaimer) I am not officially a repo maintainer. Thanks!

**Comment by anthonydahanne** [comment]
this PR has been split into 2 new PRs:

* https://github.com/spring-cloud/spring-cloud-bindings/pull/99 and
* https://github.com/spring-cloud/spring-cloud-bindings/pull/100

The goal is to keep SB 2 compatibility in [branch 1.x](https://github.com/spring-cloud/spring-cloud-bindings/tree/1.x)
and allow the [main branch](https://github.com/spring-cloud/spring-cloud-bindings/tree/main) to evolve free of SB 2.x constraints.

Similar to what other `spring-cloud` projects are doing.

**Pull request closed without merging**
